JENA-1578

GregAlbiston opened the pull request:
- Subclass ValueReplacement and methods added to allow substitution of varNames with collections of RDFNodes. This is used in SPARQL queries to provide inline data with the VALUES keyword.
- Supports multiple values for a single variable, sets of values for multiple variables and multiple sets of values for multiple values.
- Optional inclusion of parenthesis for the first two cases.
- Tests included for use cases.

ajs6f commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1740,243 @@ public String toString() {
+     * @param valuesItems
+     */
+    public void setValues(Map<String, Collection<? extends RDFNode>> valuesItems) {
+        for (String varName : valuesItems.keySet()) {
```
Just a style thing, really, but I believe this method could be just `valuesItems.forEach(this::setValues)`.

ajs6f commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1740,243 @@ public String toString() {
+    public void setValues(Map<String, Collection<? extends RDFNode>> valuesItems, Boolean isParenthesisNeeded) {
+        for (String varName : valuesItems.keySet()) {
+            Collection<? extends RDFNode> items = valuesItems.get(varName);
+            setValues(varName, items, isParenthesisNeeded);
```
Similarly, `valuesItems.forEach((varName, item)->setValues(varName, items, isParenthesisNeeded))`, and the same below.

afs commented:
@ajs6f Will take a look.

A particular thing to be careful of with ParameterizedSparqlString in general is injection attacks.

cc @rvesse

afs commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,237 @@ public String toString() {
+         */
+        private String createInsert(RDFNode item) {
+            String insert;
+            if (item.isLiteral()) {
```
Literals with language tags need treating specially. Lang tag literals have datatype rdf:langString, which should not be printed, and the @lang must be printed. 

Literal lexical forms need escaping for any ", newline and tabs in the string.

Other forms could be made to be more natural : strings without "^^xsd:string", integers as `123` etc.

The good news is that there is code to help at the node level so the literal and URI cases can be replaced by `FmtUtils.stringForNode(rdfNode.asNode(), (PrefixMapping)null)` - must use the "(PrefixMapping)null" form otherwise a helpful (for display) prefix mapping is used. This is the code used by the SPARQLserializer.

This includes the delimiters e.g `<...>` or string `"..."`.

The URI may need care in case "`...> finish URI early + something to follow`" is injected.

afs commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,237 @@ public String toString() {
+            } else if (item.isResource()) {
+                insert = "<" + item.asResource().getURI() + ">";
+            } else {
+                insert = item.asResource().getId().getLabelString();
```
Doesn't this need the "_:" put in? A test case will pin this down.

Can the label string be an injection attack with a space or newline to end the label? Probably need to restrict labels to `alpha-alphanumeric*`. (Mostly checking happens in the parsers in tokensing which is pretty heavy for this, though there may be code somewhere else that validates blank node labels.)

GregAlbiston commented:
I've made updates to try and address the comments made so far. 

- Now using _FmtUtils.stringForNode_ for conversion of RDFNode to replacement string.
- Local method _validateParameterValue_ now used as values are being set to prevent injection attack.
- Local method _validateSafeToInject_ now used when the query is being parsed to prevent injection attack. This is called each of the target variables for each relevant item. i.e. Not the varName supplied for the substitution but the variable in the VALUES clause of the query that will be evaluated.
- The need for parenthesis is now determined when the query is being parsed based on number of target variables in the VALUES clause and then, for a single target variable, the presence of parenthesis in the query.
- Updated tests, removed methods no longer required and added additional JavaDoc comments.

Thanks,

Greg

afs commented:
There is no need to use the un-parenthesed form - it is a syntactic short form and the full form can always be used. This may make the code generation easier and have less at risk of misalignment.
```
SELECT * {
       VALUES (?x) { (1) (2) (3) }
       }
```
The query formatter does much the same - if it one variable, it uses the short form, but it does not worry about the data being mis aligned:

It is the case of (illegal) `VALUES ?x { (1 2) (3) }` becoming treated as `VALUES ?x { 1 2 3 }`.

rvesse commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,250 @@ public String toString() {
+                            validateSafeToInject(command, targetVar, item.asNode());
+                        } else {
+                            //Multiple vars with items not in an ordered list. This is parsing error.
+                            throw new ARQException("Multiple VALUES variables (" + String.join(", ", targetVars) + ") being used without an ordered list of items: " + items.toString());
```
Can this error be caught sooner at the time the user makes the original `setValues()` call?

GregAlbiston replied:
This error isn't relevant anymore as the `List` is being enforced by the `setValues` method. A check is performed that the number of target vars equals the number of 'items' in a 'row' with exception thrown. The check is performed late as the 'command' is not final in the class and can be changed using `setCommandText`. Could be performed earlier but the existing approach is to wait until `toString` is called.

rvesse commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,250 @@ public String toString() {
+     * @param items
+     */
+    public void setValues(String varName, Collection<? extends RDFNode> items) {
+        items.forEach(item -> validateParameterValue(item.asNode()));
```
Later code requires that `items` be a `List` so this code should either enforce it here or otherwise sanitise `items` into a `List` at this point

GregAlbiston commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,250 @@ public String toString() {
+        return targetVars;
+    }
+
+    protected static boolean checkParenthesis(String command, String varName) {
```
Parenthesis are now always applied.

GregAlbiston commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,250 @@ public String toString() {
+            String subCmd = command.substring(0, varIndex).toLowerCase(); //Truncate the command at the varName. Lowercase to search both types of values.
+            int valuesIndex = subCmd.lastIndexOf(VALUES_KEYWORD);
+            int bracesIndex = subCmd.lastIndexOf("{");
+            String vars = command.substring(valuesIndex + VALUES_KEYWORD.length(), bracesIndex);
```
Additional checking has been added to catch missing `VALUES` or braces. The query will be unchanged if these or the valueName are missing. The existing design is that syntax errors are caught by the `Query` or `UpdateRequest` stage.

GregAlbiston commented on jena-arq/src/test/java/org/apache/jena/query/TestParameterizedSparqlString.java:
```diff
@@ -1925,4 +1926,190 @@ public void test_param_string_bug_04() {
+        //System.out.println("Res: " + res);
+        Assert.assertArrayEquals(exp, res);
+    }
+
```
Test cases added for incorrect number of values in a row for the variables.

Query that does not contain the `VALUES` keyword will now be unchanged. The existing design is that syntax errors are caught by the `Query` or `UpdateRequest` stage.

GregAlbiston commented on jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java:
```diff
@@ -1734,4 +1739,250 @@ public String toString() {
+     * @param varName
+     * @param groupedItems
+     */
+    public void setGroupedValues(String varName, Collection<List<? extends RDFNode>> groupedItems) {
```
'setGroupedValues' now renamed to 'setRowValues'. Replaced 'group' with 'row' and 'varName' with 'valueName'.
